### PR TITLE
fix issue 18456 - crt_constructor/crt_destructor segfaults if -lib

### PR DESCRIPTION
**Pull request by rainers**
ensure s.csym initialized

### CODE REVIEW

**Review by jacob-carlborg — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Comment by rainers** [comment]
I just noticed that even if no longer crashing, it still doesn't work as expected: the initializing reference is placed into a different object file inside the library, so never linked and the function is actually not called.

I'll remove auto-merge and see if a fix can be added. AFAICT auto-merge doesn't work on stable anyway ATM?

**Comment by rainers** [comment]
Works now, but 2 implementation and tests very different. Some refactorings might be desirable:
- use flags enum for `isCrtCtorDtor`
- move evaluation into semantic analysis so its usage is not limited to the glue layer

I'd rather make these changes on master, though.

File: src/dmd/glue.d [hunk 1]
```diff
@@ -1244,6 +1244,11 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
     if (fd.isExport())
         objmod.export_symbol(s, cast(uint)Para.offset);
 
+    if (fd.isCrtCtorDtor & 1)
+        objmod.setModuleCtorDtor(s, true);
+    if (fd.isCrtCtorDtor & 2)
+        objmod.setModuleCtorDtor(s, false);
```
Comment (WalterBright) [inline src/dmd/glue.d]: When using magic numbers, please document them at the definition of isCrtCtorDtor at a minimum.

File: src/dmd/toobj.d [hunk 2]
```diff
@@ -806,7 +803,7 @@ void toObjFile(Dsymbol ds, bool multiobj)
                     }
                     else if (auto f = s.isFuncDeclaration())
                     {
-                        objmod.setModuleCtorDtor(s.csym, isCtor);
+                        f.isCrtCtorDtor |= isCtor ? 1 : 2;
```
Comment (WalterBright) [inline src/dmd/toobj.d]: Why the `|=` ? Is there ever a case where 1 and 2 can both be set? Or any other bits?

File: src/dmd/glue.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/glue.d]: I see you did. Never mind.

**Review by WalterBright — DISMISSED** [review]
Fix the `|=` then it's good to go.

File: src/dmd/toobj.d (hunk 2 above)
Comment (rainers) [reply src/dmd/toobj.d]: I was a bit surprised, too, but the existing implementation allows both: https://github.com/dlang/dmd/blob/master/test/runnable/test17868b.d#L7

**Comment by thewilsonator** [comment]
Adding 72h -> merge

File: src/dmd/toobj.d (hunk 2 above)
Comment (MartinNowak) [reply src/dmd/toobj.d]: The pragma just places a function pointer in the init/fini segment, so as a low-level tool it can be used as people want.

File: src/dmd/toobj.d [hunk 3]
```diff
@@ -819,6 +816,8 @@ void toObjFile(Dsymbol ds, bool multiobj)
                 if (recurse(pd, isCtor) > 1)
                     pd.error("can only apply to a single declaration");
             }
+
+            visit(cast(AttribDeclaration)pd);
```
Comment (MartinNowak) [inline src/dmd/toobj.d]: Why did you move the call to the base class?

**Comment by MartinNowak** [comment]
This PR had quite a bad timing during the beta for such a low-level change.

File: src/dmd/toobj.d (hunk 3 above)
Comment (rainers) [reply src/dmd/toobj.d]: The flag has to be set before visiting the function generation, because the pointers in init/fini are now written when visiting the function, which might be delayed with multiobj.
The problem was that they ended up in a different object file.

**Comment by rainers** [comment]
> This PR had quite a bad timing during the beta for a low-level compiler change.

It's been lingering here quite some time, but I doubt a lot of people have used the feature as it is still undocumented.
